README
An attempt at creating a native theme for the Mac OS X port of GTK+.


ChangeLog
2009-06-03  Rob Caelers <robc@krandor.nl>

	* src/quartz-style.c (draw_check):Add drawing of GtkCellRendererToggle.

2009-06-03  Rob Caelers <robc@krandor.nl>

	* src/quartz-style.c (draw_flat_box): Fix background color of
	TreeView widget.

2009-06-03  Rob Caelers <robc@krandor.nl>

	* src/quartz-style.c (style_setup_rc_style): Fix font size for
	Gtkmm TreeView widget (widget is called gtkm__GtkTreeView and was
	not matched by style rules).

2009-06-02  Rob Caelers <robc@krandor.nl>

	* src/quartz-style.c (draw_box): Add support for the (now
	obsolete) Gtk::OptionMenu

2009-06-01  Rob Caelers <robc@krandor.nl>

	* src/quartz-style.c (draw_box): Add support for vertical progress
	bars.

2009-06-01  Rob Caelers <robc@krandor.nl>

	* .gitignore:
	* src/.gitignore: Add .git ignore files.

2008-10-30  Richard Hult  <richard@imendio.com>

	* src/quartz-style.c (is_icon_only_button, draw_box): Remove the
	heuristics that tried to decide if a square or aqua button should
	be used depending on the size. It needs to be improved before it's
	enabled again. Use square buttons for image only buttons in the
	meantime.

2008-10-29  Richard Hult  <richard@imendio.com>

	* src/quartz-style.c: Minor style cleanups.

2008-10-29  Richard Hult  <richard@imendio.com>

	* src/quartz-style.c: (draw_layout): Add missing parameters for
	area, widget, etc. Fixes missing clipping of text.

2008-10-29  Richard Hult  <richard@imendio.com>

	* src/quartz-style.c: Change get_context so it takes the window
	and handle pixmaps as well. Remove the special cases since they
	are not needed anymore. Add release_context and do the same
	special casing in there.

2008-10-29  Richard Hult  <richard@imendio.com>

	* theme/gtkrc: Change menu icon size to 16x16.

2008-10-29  Richard Hult  <richard@imendio.com>

	* src/quartz-style.c: (style_setup_rc_styles): Remove unused
	styles.

2008-10-29  Richard Hult  <richard@imendio.com>

	* src/quartz-style.c: (find_last_notebook_tab_forall),
	(draw_extension): Add notebook tab drawing.
	* theme/gtkrc: Forgot to commit those before.

2008-10-29  Richard Hult  <richard@imendio.com>

	* src/quartz-style.c: (draw_shadow): Improve the look of scrolled
	window, frame, entry. Remove broken notebook code for now, it's
	not useful in its current state.

2008-10-28  Richard Hult  <richard@imendio.com>

	* src/quartz-style.c Small whitespace cleanup.

2008-10-28  Richard Hult  <richard@imendio.com>

	* src/quartz-style.c (get_context): Add helper for getting the
	context, will be used to clip to the area when necessary soon.
	(draw_arrow): Don't draw arrows for spin buttons, they are drawn
	by the button itself.
	- Change all drawing functions to use get_context.

2008-10-28  Richard Hult  <richard@imendio.com>

	* src/quartz-style.c (style_setup_rc_styles): Handle subclasses of
	treeview as well.

2008-10-28  Richard Hult  <richard@imendio.com>

	* src/quartz-style.c (draw_handle): Implement drawing paned
	handles.

2008-10-28  Richard Hult  <richard@imendio.com>

	* src/quartz-style.c (style_setup_rc_styles): Use tooltip colors
	from the color scheme.

2008-10-28  Richard Hult  <richard@imendio.com>

	* theme/gtkrc: Add a color scheme.

2008-04-08  Richard Hult  <richard@imendio.com>

	* src/quartz-style.c: (draw_arrow), (draw_extension): Small coding
	style cleanups.

2008-04-08  Richard Hult  <richard@imendio.com>

	* src/*: More style cleanups, make all the rc style definitions
	use the macros. Add copyright information and header guards.

2008-04-07  Richard Hult  <richard@imendio.com>

	* src/quartz-rc-style.h:
	* src/quartz-style.c:
	* src/quartz-style.h:
	* src/test.c: More style fixes.

2008-04-07  Richard Hult  <richard@imendio.com>

	* src/quartz-draw.c:
	* src/quartz-style.c: Coding style fixes.

2008-04-07  Richard Hult  <richard@imendio.com>

	* src/quartz-style.c: (draw_arrow): Fix typo, makes arrowing
	drawing work.

2007-04-07  Alberto Ruiz <aruiz@gnome.org>

	* Initial commit.


src/test.c
#include <gtk/gtk.h>

int
main (int argc, char **argv)
{
  GtkWidget *window, *vbox, *entry, *button;

  gtk_init (&argc, &argv);

  window = gtk_window_new (GTK_WINDOW_TOPLEVEL);
  g_signal_connect (window,
                    "destroy",
                    G_CALLBACK (gtk_main_quit),
                    NULL);

  if (1)
    {
      vbox = gtk_vbox_new (FALSE, 12);
      gtk_container_add (GTK_CONTAINER (window), vbox);
      gtk_container_set_border_width (GTK_CONTAINER (window), 12);

      entry = gtk_entry_new ();
      gtk_box_pack_start (GTK_BOX (vbox), entry, FALSE, FALSE, 0);

      button = gtk_button_new_with_label ("Quit");
      g_signal_connect (button,
                        "clicked",
                        G_CALLBACK (gtk_main_quit),
                        NULL);
      gtk_box_pack_start (GTK_BOX (vbox), button, FALSE, FALSE, 0);
    }

  gtk_widget_show_all (window);
  gtk_main ();

  return 0;
}


autogen.sh
#!/bin/sh
# Run this to generate all the initial makefiles, etc.

PROJECT=quartz-engine

srcdir=`dirname $0`
test -z "$srcdir" && srcdir=.
DIE=0

THEDIR=`pwd`
cd $srcdir

(autoconf --version) < /dev/null > /dev/null 2>&1 || {
	echo
	echo "You must have autoconf installed to compile GTK+."
	echo "Download the appropriate package for your distribution,"
	echo "or get the source tarball at http://ftp.gnu.org/gnu/autoconf/"
	DIE=1
}

if automake-1.10 --version < /dev/null > /dev/null 2>&1 ; then
    AUTOMAKE=automake-1.10
    ACLOCAL=aclocal-1.10
elif automake-1.9 --version < /dev/null > /dev/null 2>&1 ; then
    AUTOMAKE=automake-1.9
    ACLOCAL=aclocal-1.9
elif automake-1.8 --version < /dev/null > /dev/null 2>&1 ; then
    AUTOMAKE=automake-1.8
    ACLOCAL=aclocal-1.8
elif automake-1.7 --version < /dev/null > /dev/null 2>&1 ; then
    AUTOMAKE=automake-1.7
    ACLOCAL=aclocal-1.7
else
        echo
        echo "You must have automake >=1.7.x installed to compile $PROJECT."
        echo "Install the appropriate package for your distribution,"
        echo "or get the source tarball at http://ftp.gnu.org/gnu/automake/"
        DIE=1
fi

(libtool --version) < /dev/null > /dev/null 2>&1 || {
	echo
	echo "You must have libtool installed to compile $PROJECT."
	echo "Get http://ftp.gnu.org/gnu/libtool/libtool-1.5.22.tar.gz"
	echo "(or a newer version if it is available)"
	DIE=1
}
if test "$DIE" -eq 1; then
	exit 1
fi

libtoolize --force --copy

$ACLOCAL $ACLOCAL_FLAGS
autoheader
$AUTOMAKE --add-missing
autoconf
cd $THEDIR

$srcdir/configure --enable-maintainer-mode "$@"

echo 
echo "Now type 'make' to compile $PROJECT"


src/quartz-draw.c
/* GTK+ theme engine for the Quartz backend
 *
 * Copyright (C) 2007-2008 Imendio AB
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#include <config.h>
#include <gtk/gtk.h>
#include <Carbon/Carbon.h>

#define IS_DETAIL(d,x) (d && strcmp (d, x) == 0)

/* FIXME: Fix GTK+ to export those in a quartz header file. */
CGContextRef gdk_quartz_drawable_get_context     (GdkDrawable  *drawable,
                                                  gboolean      antialias);
void         gdk_quartz_drawable_release_context (GdkDrawable  *drawable,
                                                  CGContextRef  context);

static void
quartz_measure_button (HIThemeButtonDrawInfo *draw_info,
                       gint                   width,
                       gint                   height)
{
  HIRect in_rect, out_rect;
  HIShapeRef shape;

  in_rect.origin.x = 0;
  in_rect.origin.y = 0;
  in_rect.size.width = width;
  in_rect.size.height = height;

  HIThemeGetButtonShape (&in_rect, draw_info, &shape);
  HIShapeGetBounds (shape, &out_rect);
  g_print ("Shape: %d %d, %d %d  ",
           (int) out_rect.origin.x,
           (int) out_rect.origin.y,
           (int) out_rect.size.width - (int) out_rect.origin.x,
           (int) out_rect.size.height - (int) out_rect.origin.y);

  HIThemeGetButtonBackgroundBounds (&in_rect, draw_info, &out_rect);
  g_print ("Bounds: %d %d, %d %d  ",
           (int) out_rect.origin.x,
           (int) out_rect.origin.y,
           (int) out_rect.size.width - (int) out_rect.origin.x,
           (int) out_rect.size.height - (int) out_rect.origin.y);

  HIThemeGetButtonContentBounds (&in_rect, draw_info, &out_rect);
  g_print ("Content: %d %d, %d %d\n",
           (int) out_rect.origin.x,
           (int) out_rect.origin.y,
           (int) out_rect.size.width - (int) out_rect.origin.x,
           (int) out_rect.size.height - (int) out_rect.origin.y);
}


void
quartz_draw_button (GtkStyle        *style,
                    GdkWindow       *window,
                    GtkStateType     state_type,
                    GtkShadowType    shadow_type,
                    GtkWidget       *widget,
                    const gchar     *detail,
                    ThemeButtonKind  kind,
                    gint             x,
                    gint             y,
                    gint             width,
                    gint             height)
{
  CGContextRef context;
  HIRect rect;
  HIThemeButtonDrawInfo draw_info;
  gint line_width;

  draw_info.version = 0;
  draw_info.kind = kind;
  draw_info.adornment = kThemeAdornmentNone;
  draw_info.value = kThemeButtonOff;

  if (GTK_IS_TOGGLE_BUTTON (widget) &&
      gtk_toggle_button_get_active (GTK_TOGGLE_BUTTON (widget)))
    {
      if (kind == kThemeBevelButton)
        draw_info.kind = kThemeBevelButtonInset;
      else
        draw_info.kind = kThemePushButtonInset;

      //draw_info.value = kThemeButtonOn;
    }

  if (state_type == GTK_STATE_ACTIVE ||
      (GTK_IS_TOGGLE_BUTTON (widget) &&
       gtk_toggle_button_get_active (GTK_TOGGLE_BUTTON (widget))))
    draw_info.state = kThemeStatePressed;
  else if (state_type == GTK_STATE_INSENSITIVE)
    draw_info.state = kThemeStateInactive;
  else
    draw_info.state = kThemeStateActive;

  if (GTK_WIDGET_HAS_FOCUS (widget))
    draw_info.adornment |= kThemeAdornmentFocus;

  if (IS_DETAIL (detail, "buttondefault"))
    draw_info.adornment |= kThemeAdornmentDefault;

  /* FIXME: Emulate default button pulsing. */

  gtk_widget_style_get (widget,
                        "focus-line-width", &line_width,
                        NULL);

  rect = CGRectMake (x + line_width, y + line_width,
                     width - 2 * line_width, height - 2 * line_width - 1);

  if (0)
    {
      g_print ("%d %d ", width, height);
      quartz_measure_button (&draw_info, rect.size.width, rect.size.height);
      //g_print ("%d %d\n", (int) rect.size.width, (int) rect.size.height);
    }

  context = gdk_quartz_drawable_get_context (GDK_WINDOW_OBJECT (window)->impl, FALSE);
  if (!context)
    return;

  HIThemeDrawButton (&rect,
                     &draw_info,
                     context,
                     kHIThemeOrientationNormal,
                     NULL);

  gdk_quartz_drawable_release_context (GDK_WINDOW_OBJECT (window)->impl, context);
}



src/quartz-draw.h
/* GTK+ theme engine for the Quartz backend
 *
 * Copyright (C) 2007-2008 Imendio AB
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#ifndef QUARTZ_DRAW_H
#define QUARTZ_DRAW_H

void quartz_draw_button (GtkStyle        *style,
                         GdkWindow       *window,
                         GtkStateType     state_type,
                         GtkShadowType    shadow_type,
                         GtkWidget       *widget,
                         const gchar     *detail,
                         ThemeButtonKind  kind,
                         gint             x,
                         gint             y,
                         gint             width,
                         gint             height);

#endif /* QUARTZ_DRAW_H */


src/quartz-style.h
/* GTK+ theme engine for the Quartz backend
 *
 * Copyright (C) 2007-2008 Imendio AB
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#ifndef QUARTZ_STYLE_H
#define QUARTZ_STYLE_H

#include <gtk/gtkstyle.h>

typedef struct _QuartzStyle QuartzStyle;
typedef struct _QuartzStyleClass QuartzStyleClass;

extern GType quartz_type_style;

#define QUARTZ_TYPE_STYLE              quartz_type_style
#define QUARTZ_STYLE(object)           (G_TYPE_CHECK_INSTANCE_CAST ((object), QUARTZ_TYPE_STYLE, QuartzStyle))
#define QUARTZ_STYLE_CLASS(klass)      (G_TYPE_CHECK_CLASS_CAST ((klass), QUARTZ_TYPE_STYLE, QuartzStyleClass))
#define QUARTZ_IS_STYLE(object)        (G_TYPE_CHECK_INSTANCE_TYPE ((object), QUARTZ_TYPE_STYLE))
#define QUARTZ_IS_STYLE_CLASS(klass)   (G_TYPE_CHECK_CLASS_TYPE ((klass), QUARTZ_TYPE_STYLE))
#define QUARTZ_STYLE_GET_CLASS(obj)    (G_TYPE_INSTANCE_GET_CLASS ((obj), QUARTZ_TYPE_STYLE, QuartzStyleClass))

struct _QuartzStyle
{
  GtkStyle parent_instance;
};

struct _QuartzStyleClass
{
  GtkStyleClass parent_class;
};

void quartz_style_register_type (GTypeModule *module);
void quartz_style_init          (void);

#endif /* QUARTZ_STYLE_H */


src/quartz-rc-style.h
/* GTK+ theme engine for the Quartz backend
 *
 * Copyright (C) 2007-2008 Imendio AB
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#ifndef QUARTZ_RC_STYLE_H
#define QUARTZ_RC_STYLE_H

#include <gtk/gtkrc.h>

typedef struct _QuartzRcStyle QuartzRcStyle;
typedef struct _QuartzRcStyleClass QuartzRcStyleClass;

extern GType quartz_type_rc_style;

#define QUARTZ_TYPE_RC_STYLE              quartz_type_rc_style
#define QUARTZ_RC_STYLE(object)           (G_TYPE_CHECK_INSTANCE_CAST ((object), QUARTZ_TYPE_RC_STYLE, QuartzRcStyle))
#define QUARTZ_RC_STYLE_CLASS(klass)      (G_TYPE_CHECK_CLASS_CAST ((klass), QUARTZ_TYPE_RC_STYLE, QuartzRcStyleClass))
#define QUARTZ_IS_RC_STYLE(object)        (G_TYPE_CHECK_INSTANCE_TYPE ((object), QUARTZ_TYPE_RC_STYLE))
#define QUARTZ_IS_RC_STYLE_CLASS(klass)   (G_TYPE_CHECK_CLASS_TYPE ((klass), QUARTZ_TYPE_RC_STYLE))
#define QUARTZ_RC_STYLE_GET_CLASS(obj)    (G_TYPE_INSTANCE_GET_CLASS ((obj), QUARTZ_TYPE_RC_STYLE, QuartzRcStyleClass))

struct _QuartzRcStyle
{
  GtkRcStyle parent_instance;
};

struct _QuartzRcStyleClass
{
  GtkRcStyleClass parent_class;
};

void quartz_rc_style_register_type (GTypeModule *module);

#endif /* QUARTZ_RC_STYLE_H */


src/quartz-theme-main.c
/* GTK+ theme engine for the Quartz backend
 *
 * Copyright (C) 2007-2008 Imendio AB
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#include <config.h>

#include <gmodule.h>
#include <gtk/gtk.h>

#include "quartz-style.h"
#include "quartz-rc-style.h"

G_MODULE_EXPORT void
theme_init (GTypeModule * module)
{
  quartz_rc_style_register_type (module);
  quartz_style_register_type (module);
  
  quartz_style_init ();
}

G_MODULE_EXPORT void
theme_exit (void)
{
}

G_MODULE_EXPORT GtkRcStyle *
theme_create_rc_style (void)
{
  return g_object_new (QUARTZ_TYPE_RC_STYLE, NULL);
}

G_MODULE_EXPORT const gchar *
g_module_check_init (GModule * module)
{
  return gtk_check_version (2, 10, 0);
}


src/quartz-rc-style.c
/* GTK+ theme engine for the Quartz backend
 *
 * Copyright (C) 2007-2008 Imendio AB
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#include <config.h>

#include "quartz-style.h"
#include "quartz-rc-style.h"

static void      quartz_rc_style_init         (QuartzRcStyle      *style);
static void      quartz_rc_style_class_init   (QuartzRcStyleClass *klass);
static GtkStyle *quartz_rc_style_create_style (GtkRcStyle         *rc_style);

static GtkRcStyleClass *parent_class;

GType quartz_type_rc_style = 0;

void
quartz_rc_style_register_type (GTypeModule * module)
{
    static const GTypeInfo object_info = {
	sizeof (QuartzRcStyleClass),
	(GBaseInitFunc) NULL,
	(GBaseFinalizeFunc) NULL,
	(GClassInitFunc) quartz_rc_style_class_init,
	NULL,			/* class_finalize */
	NULL,			/* class_data */
	sizeof (QuartzRcStyle),
	0,			/* n_preallocs */
	(GInstanceInitFunc) quartz_rc_style_init,
    };

    quartz_type_rc_style = g_type_module_register_type (module,
						     GTK_TYPE_RC_STYLE,
						     "QuartzRcStyle",
						     &object_info, 0);
}

static void
quartz_rc_style_init (QuartzRcStyle * style)
{
}

static void
quartz_rc_style_class_init (QuartzRcStyleClass * klass)
{
    GtkRcStyleClass *rc_style_class = GTK_RC_STYLE_CLASS (klass);

    parent_class = g_type_class_peek_parent (klass);

    rc_style_class->create_style = quartz_rc_style_create_style;
}

static GtkStyle *
quartz_rc_style_create_style (GtkRcStyle * rc_style)
{
    return g_object_new (QUARTZ_TYPE_STYLE, NULL);
}
